### PR TITLE
Add tools for axon filtering

### PR DESCRIPTION
**Pull request by Stoyan-I-A**
## Checklist

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've added relevant tests for my contribution
- [x] I've updated the documentation and/or added correct docstrings
- [x] I've assigned a reviewer
- [x] I've consulted [ADS's internal developer documentation](https://github.com/neuropoly/axondeepseg/wiki/Guidelines-for-contribution) to ensure my contribution is in line with any relevant design decisions

## Description
The purpose of this PR is to add the functions for removing axonmyelin objects from masks.
It does not add a new feature directly, but the functions can easily be used in Jupyter notebooks or other scripts and functions.

## Linked issues
Related to #628 but does not implement the feature.

### CODE REVIEW

**Comment by Stoyan-I-A** [comment]
@hermancollin I added you as a reviewer since you already played with these new functions.

**Comment by Stoyan-I-A** [comment]
There's a test that's failing, see [this comment](https://github.com/axondeepseg/axondeepseg/issues/647#issuecomment-1166700658) for more info.

**Comment by mathieuboudreau** [comment]
> There's a test that's failing, see [this comment](https://github.com/axondeepseg/axondeepseg/issues/647#issuecomment-1166700658) for more info.

That's odd that the tests in the PR passed without the files, but after merging to master they are failing. Typically PRs test for that (running tests for the scenario after merging).

**Comment by Stoyan-I-A** [comment]
@mathieuboudreau The tests are passing on master because they are NOT using the [latest release](https://github.com/axondeepseg/data-testing/releases/tag/r20220607a) which contains a file for the reference morphometrics that has makes the tests fail.

Since I created a pre-release for this PR and use it, it had to include this new file that was overlooked  in PR #650.

**Comment by Stoyan-I-A** [comment]
The test is now passing. Thank you @hermancollin!

File: AxonDeepSeg/postprocessing.py
Comment (hermancollin) [inline AxonDeepSeg/postprocessing.py:204]: this function is a nice touch

**Review by hermancollin — DISMISSED** [review]
Tested with the image from #659. It works well for simple cases. For more complex ones it also does a decent job, but the output is sometimes unpredictable. Removing the axon at [0,0] on this image changed the myelin mask like this:

00_removed

The axon masks on the other hand are pretty much always perfect. All in all, this will be very useful for future developments related to #628. Thanks a lot @Stoyan-I-A for the work you put into this.

**Comment by Stoyan-I-A** [comment]
> For more complex ones it also does a decent job, but the output is sometimes unpredictable. Removing the axon at [0,0] on this image changed the myelin mask like this:

Are you talking about the weird edges of the axons adjacent to the one removed or the fact that other myelin bits in the middle of the image are removed 😳?
I can explain the 1st scenario but the second one seems like a bug.

**Comment by hermancollin** [comment]
> > For more complex ones it also does a decent job, but the output is sometimes unpredictable. Removing the axon at [0,0] on this image changed the myelin mask like this:
> 
> Are you talking about the weird edges of myelin of axons adjacent to the one removed or the fact that other myelin bits in the middle of the image are removed 😳? I can explain the 1st scenario but the second one seems like a bug.

Yes I'm talking about the pieces in the middle and on the borders far from (0,0). To be fair, I purposefully chose a challenging image.

**Comment by Stoyan-I-A** [comment]
I took a closer look at the image in #659. The pieces of myelin that disappear are myelin chunks that are not attached to an axon. Therefore, they are not detected by the watershed because we use axon centroids as seed points.

That being said, I have an idea on how to solve this, so don't merge the change just yet!

**Comment by Stoyan-I-A** [comment]
I changed the logic a little bit. It shouldn't remove the loose myelin bits anymore

**Comment by hermancollin** [comment]
@Stoyan-I-A thank you for correcting this. For future reference, before we merge this, could you explain how you fixed the loose myelin issue? From what I can see, you removed these loose myelin bits from the axonmyelin mask before performing the final logical AND but I'm not sure I fully understand the operation.

Removing the axon at `0,0` (myelin mask) now yields:

00_removed_20220711

**Review by hermancollin — APPROVED** [review]

**Comment by Stoyan-I-A** [comment]
I tried to clarify the process used to remove the axons in my new commit. Unfortunately, this means you'll have to approuve it again.

What happened before my patch is that I performed a flood fill of value 0 at the location of the axons to remove on the watershed segmentation. That being said, the loose myelin objects also had a value of 0 on the watershed segmentation because they weren't attached to an axon because the seed points of the watershed are on the centroids of axon objects. So they were identified as background right when applying the watershed.

Now instead of applying a value of 0 on the axons to remove, I apply a value of 65535 (max value of uint16) on them. I can then remove the axon and myelin parts which correspond to regions with a value of 65535 on the watershed segmentation after the flood fill.

**Review by hermancollin — APPROVED** [review]
Great! Thanks for the clarification.

**Pull request merged**
